Modules/unicodedata.c

```diff
@@ -1439,43 +1821,85 @@ this database is based on the UnicodeData.txt file version\n\
 The module uses the same names and symbols as defined by the\n\
 UnicodeData File Format " UNIDATA_VERSION ".");
 
-static struct PyModuleDef unicodedatamodule = {
-        PyModuleDef_HEAD_INIT,
-        "unicodedata",
-        unicodedata_docstring,
-        -1,
-        unicodedata_functions,
-        NULL,
-        NULL,
-        NULL,
-        NULL
-};
+static int
+unicodedata_traverse(PyObject *module, visitproc visit, void *arg)
+{
+    unicodedata_state *state = unicodedata_get_state(module);
+    Py_VISIT(state->ucd_type);
+    return 0;
+}
 
-PyMODINIT_FUNC
-PyInit_unicodedata(void)
+static int
+unicodedata_clear(PyObject *module)
 {
-    PyObject *m, *v;
+    unicodedata_state *state = unicodedata_get_state(module);
+    Py_CLEAR(state->ucd_type);
+    return 0;
+}
 
-    Py_SET_TYPE(&UCD_Type, &PyType_Type);
+static void
+unicodedata_free(void *module)
+{
+    unicodedata_clear((PyObject *)module);
+}
 
-    m = PyModule_Create(&unicodedatamodule);
-    if (!m)
-        return NULL;
+static int unicodedata_exec(PyObject *m)
+{
+    unicodedata_state *state = unicodedata_get_state(m);
+    state->ucd_type = (PyTypeObject *)PyType_FromModuleAndSpec(
+        m, &unicodedata_ucd_type_spec, NULL);
+    if (state->ucd_type == NULL) {
+        return -1;
+    }
 
-    PyModule_AddStringConstant(m, "unidata_version", UNIDATA_VERSION);
-    Py_INCREF(&UCD_Type);
-    PyModule_AddObject(m, "UCD", (PyObject*)&UCD_Type);
+    if (PyModule_AddStringConstant(m, "unidata_version", UNIDATA_VERSION) < 0) {
+        return -1;
+    }
 
     /* Previous versions */
-    v = new_previous_version("3.2.0", get_change_3_2_0, normalization_3_2_0);
-    if (v != NULL)
-        PyModule_AddObject(m, "ucd_3_2_0", v);
+    PyObject *v = new_previous_version(
+        state, "3.2.0", get_change_3_2_0, normalization_3_2_0);
+    if (v == NULL) {
+        return -1;
+    }
+    if (PyModule_AddObject(m, "ucd_3_2_0", v) < 0) {
+        Py_DECREF(v);
+        return -1;
+    }
 
     /* Export C API */
     v = PyCapsule_New((void *)&hashAPI, PyUnicodeData_CAPSULE_NAME, NULL);
```

Comment:
I'm not sure if passing state=NULL and self=NULL to internal functions through the PyCapsule C API is correct.

Reply:
@vstinner what do you mean by correct?  Are you concerned I changed the behavior or are you concerned about maintainability?

Reply:
I'm worried about correctness. For example, I don't know what is the behavior when PyCapsule_New() is called with the same name twice. Does PyCapsule_New() override the old values? Which behavior do we want?
https://bugs.python.org/issue1635741#msg376555

Moreover, currently, PyCapsule_New() API pass NULL as the module parameter, so it's possible to retrieve the module state yet.

Reply:
@vstinner I tried PyCapsule_SetContext and GetContext and it will not help, this is mainly when a destructor is defined.

What do you mean by module parameter in PyCapsule_New?